cpp/src/arrow/flight/sql/client.h

```diff
@@ -329,6 +329,25 @@ class ARROW_FLIGHT_SQL_EXPORT FlightSqlClient {
   /// \param[in] info         The FlightInfo of the query to cancel.
   ::arrow::Result<CancelResult> CancelQuery(const FlightCallOptions& options,
                                             const FlightInfo& info);
+   
+  /// \brief Sets session options.
+  ///
+  /// \param[in] options            RPC-layer hints for this call.
+  /// \param[in] session_options    The session options to set.
+  ::arrow::Result<std::vector<SetSessionOptionResult>> SetSessionOptions(
+      const FlightCallOptions& options,
+      const std::vector<SessionOption>& session_options);
```

Comment:
Originally we were not going to but given we're necessarily going to be setting the desired Resource-Path session option value /after/ connecting I don't see that it's going to be immutable in any way.  If we start thinking in terms of it being updated we do need to ensure that it's captured at call time by the FlightProducer and can't change from that call handing code's perspective so that there isn't a race if e.g. SetSessionOptions is called again while a query is busy executing or whatnot (i.e. snapshot the session state for the FlightProducer's handler's perspective or read/copy the values at handler init or whatever).